faculty/clients/serveragent.py

```diff
@@ -0,0 +1,344 @@
+        The status of the execution.
+    environments : List[EnvironmentExecution]
+        list of EnvironmentExecution objects for each environment applied.
+    started_at : Optional[datetime]
+        startime of the execution.
+    finished_at : Optional[datetime]
+        endtime of the execution.
+    """
+
+    id = attrib()
+    status = attrib()
+    environments = attrib()
+    started_at = attrib()
+    finished_at = attrib()
+
+
+@attrs
+class EnvironmentExecution(object):
+    """An environment executed on a server.
+    
+    Parameters
+    ----------
+    id : uuid.UUID
+        The ID of the environment.
+    steps : List[EnvironmentExecutionStep]
+        List of EnvironmentExecutionStep objects.
+    """
+
+    id = attrib()
+    steps = attrib()
+
+
+@attrs
+class EnvironmentExecutionStep(object):
+    """A single environment execution step on a server.
+    
+    Parameters
+    ----------
+    id : uuid.UUID
+        The ID of the environment step.
+    command : List[str]
+        The command executed by the step.
+    status : enum.Enum
+        The status of the step execution.
+    started_at : Optional[datetime]
+        The start time of the execution step.
+    finished_at : Optional[datetime]
+        The finish time of the execution step.
+    """
+
+    id = attrib()
+    command = attrib()
+    status = attrib()
+    started_at = attrib()
+    finished_at = attrib()
+
+
+@attrs
+class EnvironmentExecutionStepLog(object):
+    """The log for an environment execution step.
+    
+    Parameters
+    ----------
+    line_number : int
+        The line number within the step log.
+    content : str
+        The content of the step log.
+    """
+
+    line_number = attrib()
+    content = attrib()
+
+
+
+@attrs
+class ServerResources(object):
+    """Information about current server resource usage.
+    
+    Parameters
+    ----------
+    milli_cpus : CpuUsage
+        current CPU utilisation.
+    memory_mb : MemoryUsage
+        current memory utilisation. 
+    """
+
+    milli_cpus = attrib()
+    memory_mb = attrib()
+
+
+@attrs
+class CpuUsage(object):
+    """Current CPU usage on a server.
+    
+    Parameters
+    ----------
+    total : int
+        total CPU resource.
+    used : int
+        currently utilised CPU resource.
+    """
+
+    total = attrib()
+    used = attrib()
+
+
+@attrs
+class MemoryUsage(object):
+    """Current Memory usage on a server.
+    
+    Parameters
+    ----------
+    total : float
+        total memory resource.
+    used : float
+        current utilised memory resource.
+    cache : float
+        current cache memory usage.
+    rss: float
+        resident set size allocation. 
+    """
+
+    total = attrib()
+    used = attrib()
+    cache = attrib()
+    rss = attrib()
+
+
+class ServerAgentClient(BaseClient):
+    """Client for the Faculty server events service.
+
+    Build this client with a session directly.
+
+    >>> client = ServerAgenClient(url, session)
```

Comment:
How about a slightly more complete example in a "Usage" section, e.g.:

```suggestion
    Usage
    -----
    
    This client needs to be constructed manually with a Faculty session and
    the URL of the agent of the server you wish to access:
    
    >>> import faculty
    >>> import faculty.session
    >>> from faculty.clients import ServerAgentClient
    >>> 
    >>> server_client = faculty.client("server")
    >>> server = server_client.get(...)  # TODO: put representative example arguments here
    >>> url = # TODO: complete line to extract hound URL
    >>> 
    >>> session = faculty.session.get_session()
    >>> server_agent_client = ServerAgentClient(url, session)
```